### PR TITLE
Introduce DT-based OpenAMP IPC service backend

### LINKED ISSUE #37130: Add multi-core IPC service
With #36120 now merged we can start discussing how to extend the IPC service to support multi-core (>2) IPC.

The idea is that  the IPC Service backend implementation should be able to create several RPMsg instances based on DT information for IPC (shared memory regions, target CPU, IPM instances and mechanism, etc.. ) during system initialization. The API should allow the user to use any of the instances created without worrying about hardware details but only defining `target_cpu` and `endpoint_name`.

Tagging the relevant people: @maje-emb @hubertmis @carlescufi

### PR DESCRIPTION
**Pull request by carlocaione**
Introduce the new DT-based OpenAMP IPC service backend. Using this new backend we can now register OpenAMP (RPMsg with static VRINGs) instances in the DT and use the IPC service APIs on these instances to register endpoints and send data through.

For example in the DT we can have something like this (on the primary core):
```
       reserved-memory {                                                           
                sram_ipc0: memory@20070000 {
                        reg = <0x20070000 0x8000>;
                };   
        };

        soc {
                peripheral@50000000 {
                        mbox: mbox@2a000 {
                                compatible = "nordic,mbox-nrf-ipc";
                                reg = <0x2a000 0x1000>;
                                tx-mask = <0x0000ffff>;
                                rx-mask = <0x0000ffff>;
                                interrupts = <42 NRF_DEFAULT_IRQ_PRIORITY>;
                                #mbox-cells = <1>;
                                status = "okay";
                        };
                };
        };

        ipc0: ipc0 {
                compatible = "subsys,ipc-openamp-static-vrings";
                shm = <&sram_ipc0>;
                mboxes = <&mbox 0>, <&mbox 1>;
                mbox-names = "tx", "rx";
                role = "primary";
                status = "okay";
        };
```
And then we can use `ipc0` to create endpoints and send data to a secondary core:
```
static struct ipc_ept_cfg ipc0_ept_cfg = { 
    .name = "ipc0",
    .cb = {
        .bound    = ipc0_ept_bound,
        .received = ipc0_ept_recv,
    };
};

struct ipc_ept ipc0_ept;

ipc0_instance = DEVICE_DT_GET(DT_NODELABEL(ipc0));
ipc_service_open_instance(ipc0_instance);
ipc_service_register_endpoint(ipc0_instance, &ipc0_ept, &ipc0_ept_cfg);
ipc_service_send(&ipc0_ept, &message, sizeof(message));
```

The backend supports multiple instances and multiple endpoints and it uses MBOX controllers.

### CODE REVIEW

**Comment by arnopo** [comment]
Hi @carlocaione 

The PR It is quite confusing for me.  the introduction of the multi instance using DT is nice. But With this PR i can see 2 multi instance backend. 
Do you plan to remove the legacy multi-instance one?
If not the case, something seems missing in commits/documentation to explain criteria to chose one backend rather than the other.

**Comment by carlocaione** [comment]
> Do you plan to remove the legacy multi-instance one?

Yes, that will be removed as soon as this is merged.

File: dts/bindings/ipc/subsys,ipc-openamp-static-vrings.yaml
Comment (hubertmis) [inline dts/bindings/ipc/subsys,ipc-openamp-static-vrings.yaml:25]: It looks the backend requires at least `tx` and `rx` mailboxes described. Should we add it to the description?

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings_mi.c
Comment (hubertmis) [inline subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings_mi.c:103]: nit: simple comparison would be more efficient
```suggestion
		if (rpmsg_instance->endpoint[i].name[0] == '\0') {
```

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (hubertmis) [inline subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:334]: It looks one of the threads can get `-EALREADY` before the other thread finishes initialization of given instance. Should we take care of such possible race condition?

**Review by hubertmis — CHANGES_REQUESTED** [review]
Overall looks good :+1: 
A few nits and a potential race condition spotted.

File: dts/bindings/ipc/subsys,ipc-openamp-static-vrings.yaml
Comment (arnopo) [reply dts/bindings/ipc/subsys,ipc-openamp-static-vrings.yaml:25]: Is possible to define twice the same mailbox? 
Regarding https://github.com/zephyrproject-rtos/zephyr/pull/38322
The same mailbox is used in RX and TX...

File: include/ipc/ipc_service_backend.h
Comment (carlocaione) [reply include/ipc/ipc_service_backend.h:32]: ACK

File: subsys/ipc/ipc_service/ipc_service.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/ipc_service.c:28]: No strong preference but `!backend->open_instance` is not always a bug. I'm thinking for example of backends that do not need to have an open function at all, they can simply to not implement it.

File: include/ipc/ipc_rpmsg.h
Comment (carlocaione) [reply include/ipc/ipc_rpmsg.h:12]: ACK

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings_mi.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings_mi.c:302]: Yes. I'm basically using the `ept->name` to cache the endpoint name when the secondary core is registering the endpoint before the primary core.

Details following.

When the secondary core registers an endpoint that is still unknown to the primary core, the backend code is caching this value into `ept->name` in: https://github.com/zephyrproject-rtos/zephyr/blob/0c095a0b67c653ad488ce8aa0052fa63e79cce31/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L162

When the primary core has the chance to register the endpoint with the same name, we can match the cached value in:
https://github.com/zephyrproject-rtos/zephyr/blob/0c095a0b67c653ad488ce8aa0052fa63e79cce31/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L278

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:73]: right. We should probably just forbid endpoints with an empty name.

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:98]: As said before I think it should be avoided having endpoints without a name.

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:183]: Yes, but TBH not my idea :) it was stolen from the legacy backend.

I tried to highlight this process with the huge comment in https://github.com/zephyrproject-rtos/zephyr/blob/0c095a0b67c653ad488ce8aa0052fa63e79cce31/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.h#L12-L78

but I'll try to clarify this by adding a new comment as you proposed.

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:321]: Agree

File: samples/subsys/ipc/ipc_service/static_vrings/remote/src/main.c
Comment (carlocaione) [reply samples/subsys/ipc/ipc_service/static_vrings/remote/src/main.c:155]: That's actually done: the `ipc0A_ept` endpoint is registered on the same `ipc0_instance`.

In the sample we have: `ipc0A_ept` and `ipc0B_ept` registered on the `ipc0` instance, and `ipc1_ept` endpoint registered on the `ipc1` instance.

File: dts/bindings/ipc/subsys,ipc-openamp-static-vrings.yaml
Comment (carlocaione) [reply dts/bindings/ipc/subsys,ipc-openamp-static-vrings.yaml:16]: Oooh, thank you for pointing this out. I didn't know that there was a proposed terminology already. I'll fix this to be consistent with OpenAMP.

File: dts/bindings/ipc/subsys,ipc-openamp-static-vrings.yaml
Comment (carlocaione) [reply dts/bindings/ipc/subsys,ipc-openamp-static-vrings.yaml:12]: Yes, I'll also match what has been proposed in https://github.com/OpenAMP/open-amp/issues/312

File: dts/bindings/vendor-prefixes.txt
Comment (carlocaione) [reply dts/bindings/vendor-prefixes.txt:575]: uhm, this was actually proposed by @mbolivar-nordic. No strong preference here.

File: dts/bindings/ipc/subsys,ipc-openamp-static-vrings.yaml
Comment (carlocaione) [reply dts/bindings/ipc/subsys,ipc-openamp-static-vrings.yaml:25]: > It looks the backend requires at least tx and rx mailboxes described. Should we add it to the description?

Yes, I'll add it.

> Is possible to define twice the same mailbox?

yes, you can use the same mailbox and you can even use the same channel for RX and TX if your hardware (and driver) supports it.

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings_mi.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings_mi.c:103]: oh, this is nice :+1:

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:334]: uhm, unlikely but not impossible. Two possible solutions here:
1. Implementing a state machine `READY->BUSY->INITED` for the instance but this means that we should explicitly busy-wait on the `open()` function when getting `BUSY`
2. Protecting the whole `open` code with a mutex / spinlock

File: subsys/ipc/ipc_service/ipc_service.c
Comment (arnopo) [reply subsys/ipc/ipc_service/ipc_service.c:28]: but in this case the ipc_service_open_instance would be backend dependent ( return 0 or -EIO). So API would be not generic.

File: samples/subsys/ipc/ipc_service/static_vrings/remote/src/main.c
Comment (arnopo) [reply samples/subsys/ipc/ipc_service/static_vrings/remote/src/main.c:155]: My apologies, I missed it

File: dts/bindings/ipc/subsys,ipc-openamp-static-vrings.yaml
Comment (carlocaione) [reply dts/bindings/ipc/subsys,ipc-openamp-static-vrings.yaml]: Moved to `host` and `remote`

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:377]: Implemented a `READY->BUSY->INITED` state machine

File: dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml
Comment (mbolivar-nordic) [inline dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml:6]: This compatible doesn't match the one in the commit message

**Review by mbolivar-nordic — APPROVED** [review]
For the binding

File: dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml
Comment (carlocaione) [reply dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml:6]: Oooh right, it was changed `s/subsys/zephyr`. I'll fix that.

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (hubertmis) [inline subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:333]: I'm thinking what happens if `register_ept()` is called in host first and right after on remote, so that remote receives `ns_bind_cb()` while the mutex is locked.

Remote code would see that `ept_cached == false`, it would cache the endpoint and unlock the mutex. Unlocking the mutex would trigger `ns_bind_cb()` handler which would call `rpmsg_create_ept()`, because the endpoint would be already cached. And right after that, `ipc_rpmsg_register_ept()` is called, which also calls `rpmsg_create_ept()` for the same endpoint.

Is it OK to call `rpmsg_create_ept()` twice for the same endpoint? What if both `register_ept()` and `ns_bind_cb()` are executed from threads with the same priority? Is `rpmsg_create_ept()` reentrant?

File: subsys/ipc/ipc_service/ipc_service.c
Comment (hubertmis) [inline subsys/ipc/ipc_service/ipc_service.c:33]: What if asserts are disabled for given build? I think we should handle `if (!backend->open_instance) return SOME_ERROR;` alongside `__ASSERT`.

**Review by hubertmis — CHANGES_REQUESTED** [review]

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:333]: > remote receives ns_bind_cb() while the mutex is locked.

This is not going to happen because `ns_bind_cb()` is only called on host (see huge comment in `subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.h`). 

More in detail the `ns_bind_cb()` callback is called **only** on the host side when on the remote side `rpmsg_create_ept()` is called with destination `RPMSG_ADDR_ANY` in  https://github.com/zephyrproject-rtos/zephyr/blob/8fbbf2baefadbf4a3d08e19a7f60de1b989dc36e/subsys/ipc/ipc_service/lib/ipc_rpmsg.c#L54-L57 during the endpoint registration.

This call is triggering the `ns_bind_cb()` on the host side. On the host side we call indeed `rpmsg_create_ept()` in https://github.com/zephyrproject-rtos/zephyr/blob/32812e7b62e1fb88c471f773332a8f274b115e6e/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L121-L122 but this time we are specifying the destination address of the remote endpoint so this is not triggering anything on the remote side (see https://github.com/OpenAMP/open-amp/blob/master/lib/rpmsg/rpmsg.c#L301-L303)

File: subsys/ipc/ipc_service/ipc_service.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/ipc_service.c:33]: Yes, I'm a bit conflicted about this considering what @arnopo said in https://github.com/zephyrproject-rtos/zephyr/pull/40365#discussion_r751501726

Comment (arnopo) [reply subsys/ipc/ipc_service/ipc_service.c:33]: Perhaps following code would be a compromize
```
if( !backend->open_instance)
     return 0;

return backend->open_instance(instance); 
```
The most important is to not return an error if not defined

File: dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml
Comment (arnopo) [inline dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml:19]: regarding https://github.com/zephyrproject-rtos/zephyr/blob/main/include/linker/devicetree_reserved.h#L121 and Linux DT implementation , `memory-region` seems to me more generic than `shm`

**Review by arnopo — CHANGES_REQUESTED** [review]
No time yet to review the sample in details...
I have few additional comments but the DT approach seems to me the right approach  to address multi-instance and multi core, but also to extend in middle/long term for some other virtio implementations.

File: include/ipc/ipc_service.h
Comment (carlocaione) [reply include/ipc/ipc_service.h:121]: I marked this as optional because some backend could decide to `open()` the instance at `init()` time (that is in the `.init` hook function of the device). This is something that I tried to do before discovering that the stack size available at `init` time is too small to accomodate for the `rpmsg_init_vdev`, so I had to migrate to a proper `open()` function.

File: include/ipc/ipc_static_vrings.h
Comment (carlocaione) [reply include/ipc/ipc_static_vrings.h:28]: Yes, this is a really good idea.

File: dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml
Comment (carlocaione) [reply dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml:10]: WOW, this is a really great new, thank you for pointing this out. I'll definitely try to align the two works.

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:77]: Yes, it is. The goal of this function is picking up an empty endpoint slot to fill. An empty endpoint slot is recognized by having an empty name.

In general it works this way: if the name is empty then the endpoint is available and it can be filled in two cases:
1. When the remote endpoint is advertising a new endpoint that it is still unknown to the host, see: https://github.com/zephyrproject-rtos/zephyr/blob/32812e7b62e1fb88c471f773332a8f274b115e6e/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L159-L166
2. When the host is registering a new endpoint that it is still not advertised by the remote endpoint. In this case we cache the name and wait for the advertising from the remote, see:
https://github.com/zephyrproject-rtos/zephyr/blob/32812e7b62e1fb88c471f773332a8f274b115e6e/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L325-L329

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:218]: ACK

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:235]: Uhm, actually `conf == NULL` should never be possible.

File: dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml
Comment (carlocaione) [reply dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml:19]: good idea

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:156]: no, it's because it's not useful to keep that locked (AFAICT). The mutex is basically protecting what `get_ept()` is returning (the endpoint), from concurrent access between `ns_bind_cb()` and `register_ept()`.

I don't think that having the mutex locked for `advertise_ept()` is useful (but I could be wrong).

File: include/ipc/ipc_service.h
Comment (arnopo) [reply include/ipc/ipc_service.h:121]: Thanks for the clarification. What about:
- make `ipc_service_open_instance` usage mandatory
- `backend->open_instance`  would be optional depending on backend
- 'ipc_service_open_instance' just do nothing if `backend->open_instance`  is NULL
This offer a "stable" API for user with possibility to implement or not the backend open instance

File: dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml
Comment (arnopo) [reply dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml:10]: The linux maintainers still have to be convinced, so this will probably change a bit.  The name of the `virtio` node might change, as well as the `virtio,id` fields and the `compatible` value.
So perhaps don't lost your time to align both. Updating the `shm` to `memory-region` is probably enough for now.

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (arnopo) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:156]: no race condition on `rpmsg_ept->bound` between `advertise_ept` and `ept_cb` ?

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (arnopo) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:77]: > Yes, it is. The goal of this function is picking up an empty endpoint slot to fill. An empty endpoint slot is recognized by having an empty name.
Understood now , thanks for the clarification 

> 
> In general it works this way: if the name is empty then the endpoint is available and it can be filled in two cases:
> 
> 1. When the remote endpoint is advertising a new endpoint that it is still unknown to the host, see: https://github.com/zephyrproject-rtos/zephyr/blob/32812e7b62e1fb88c471f773332a8f274b115e6e/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L159-L166

`name == null` should not occur on a name service annoucement. What about treat it as an error?


> 2. When the host is registering a new endpoint that it is still not advertised by the remote endpoint. In this case we cache the name and wait for the advertising from the remote, see:
>    https://github.com/zephyrproject-rtos/zephyr/blob/32812e7b62e1fb88c471f773332a8f274b115e6e/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L325-L329

Same here `cfg->name` can not be null as tested before in the function

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (arnopo) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:235]: Right, except race condition or memory corruption. And in this case even if `conf` is not null, `conf->mbox_tx `can.
if assert is not recommended, an error message could help to debug...
A niminum could you add test on `conf->mbox_tx `?

File: samples/subsys/ipc/ipc_service/static_vrings/remote/src/main.c
Comment (arnopo) [inline samples/subsys/ipc/ipc_service/static_vrings/remote/src/main.c:138]: NET? 
What about REMOTE instead of NET and HOST instead of APP ( in static_vrings/src/main.c)
https://github.com/zephyrproject-rtos/zephyr/blob/32812e7b62e1fb88c471f773332a8f274b115e6e/samples/subsys/ipc/ipc_service/static_vrings/src/main.c#L144

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (hubertmis) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:333]: So, it means that `if (ept_cached)` from line 317 can be true only in the host role, right? Maybe we could add an assertion in there to check the role. It would help following the code.

I think I finally understood how the code flow works here. I'm not sure how to improve readability. Maybe comment for `ns_bind_cb()` and `advertise_ept()` that these functions are called only in the host role?

**Review by hubertmis — APPROVED** [review]

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:77]: ACK

File: subsys/ipc/ipc_service/ipc_service.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/ipc_service.c:33]: ACK

File: include/ipc/ipc_service.h
Comment (carlocaione) [reply include/ipc/ipc_service.h:121]: OK

File: dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml
Comment (carlocaione) [reply dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml:10]: ACK

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:156]: Uhm, I don't think so. `->bound` in `advertise_ept` is exclusively called by the host (either from `ns_bind_cb` or from `register_ept`) while `->bound` in `ept_cb` is only called on the remote. I'll add a comment trying to explain better how this works.

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c
Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:235]: sure thing.

File: samples/subsys/ipc/ipc_service/static_vrings/remote/src/main.c
Comment (carlocaione) [reply samples/subsys/ipc/ipc_service/static_vrings/remote/src/main.c:138]: Yup, no prob.

**Comment by carlocaione** [comment]
> So, it means that if (ept_cached) from line 317 can be true only in the host role, right?

Correct

> Maybe we could add an assertion in there to check the role. It would help following the code.

I have a better idea ;) stay tuned.

> I think I finally understood how the code flow works here. I'm not sure how to improve readability. Maybe comment for ns_bind_cb() and advertise_ept() that these functions are called only in the host role?

Yes, I'll add another huge comment :)

**Comment by carlocaione** [comment]
@hubertmis @arnopo I added this comment that maybe can help a bit with understanding what's going on:
https://github.com/zephyrproject-rtos/zephyr/blob/b40846b4604d767055d3a494ce1699f70200af88/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.h#L82-L131

**Comment by hubertmis** [comment]
Splitting `register_ept` to `register_ept_on_host()` and `register_ept_on_remote()` helps to follow the code. The only missing piece of puzzle for me is clear information that only remote registers endpoint with dst addr ANY, what triggers `ns_bind_cb` on the host.

I think it would be easier to follow if all or none of the calls to rpmsg were abstracted in `subsys/ipc/ipc_service/lib/ipc_rpmsg.c`. Right now part of the calls goes through this helper module, and another part is called directly from the backend.

**Comment by carlocaione** [comment]
> Splitting `register_ept` to `register_ept_on_host()` and `register_ept_on_remote()` helps to follow the code. The only missing piece of puzzle for me is clear information that only remote registers endpoint with dst addr ANY, what triggers `ns_bind_cb` on the host.

I agree. I'll add a comment for that.

> I think it would be easier to follow if all or none of the calls to rpmsg were abstracted in `subsys/ipc/ipc_service/lib/ipc_rpmsg.c`. Right now part of the calls goes through this helper module, and another part is called directly from the backend.

Yes, unfortunately the `ns_bind_cb` call has been carved out from the library to allow the caching mechanism in this backend.

**Review by arnopo — APPROVED** [review]
LGTM,
Thanks @carlocaione for this very nice redesign!

**Pull request merged**
